File: app/src/main/java/me/sunzheng/mana/home/search/SearchResultAdapter.java
package me.sunzheng.mana.home.search;

import android.app.Activity;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.util.List;

import me.sunzheng.mana.BangumiDetailsActivity;
import me.sunzheng.mana.R;
import me.sunzheng.mana.core.BangumiModel;
import me.sunzheng.mana.utils.LanguageSwitchUtils;

/**
 * Created by Sun on 2017/6/20.
 */

public class SearchResultAdapter extends RecyclerView.Adapter<SearchResultAdapter.ViewHolder> {
    List<BangumiModel> mValues;

    public SearchResultAdapter(List<BangumiModel> list) {
        this.mValues = list;
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_onairfragment, parent, false);
        return new ViewHolder(view);
    }


    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                BangumiDetailsActivity.newInstance((Activity) v.getContext(), mValues.get(position).getId().toString(), mValues.get(position).getImage(), LanguageSwitchUtils.switchLanguageToJa(holder.itemView.getContext(), mValues.get(position).getName(), mValues.get(position).getNameCn()),
                        holder.mImageView);
            }
        });
        Glide.with(holder.itemView.getContext())
                .load(mValues.get(position).getImage())
                .placeholder(new ColorDrawable(Color.parseColor(mValues.get(position).getCover_color())))
                .into(holder.mImageView);
        holder.mTitleTextView.setText(LanguageSwitchUtils.switchLanguageToJa(holder.itemView.getContext(), mValues.get(position).getName(), mValues.get(position).getNameCn()));
        holder.mSummaryTextView.setText(mValues.get(position).getSummary());
        holder.mEtcTextView.setText(mValues.get(position).getAirDate());
    }

    protected final static class ViewHolder extends RecyclerView.ViewHolder {
        public final CardView mView;
        public final TextView mTitleTextView, mSummaryTextView, mEtcTextView;
        public final ImageView mImageView;

        public ViewHolder(View view) {
            super(view);
            mView = (CardView) view;
            mTitleTextView = (TextView) view.findViewById(R.id.item_title_textview);
            mImageView = (ImageView) view.findViewById(R.id.item_album);
            mSummaryTextView = (TextView) view.findViewById(R.id.item_subtitle_textview);
            mEtcTextView = (TextView) view.findViewById(R.id.item_etc_textview);
        }
    }
}

File: app/src/main/java/me/sunzheng/mana/home/search/SearchResultAdapter.kt
package me.sunzheng.mana.home.search

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SortedList
import androidx.recyclerview.widget.SortedListAdapterCallback
import me.sunzheng.mana.R
import me.sunzheng.mana.core.net.v2.database.BangumiEntity
import me.sunzheng.mana.core.net.v2.loadUrl
import me.sunzheng.mana.utils.ArrarysResourceUtils
import me.sunzheng.mana.utils.LanguageSwitchUtils

/**
 * Created by Sun on 2017/6/20.
 */
class SearchResultAdapter(
    var values: List<BangumiEntity>,
    var onItemClickListener: ((view: View, position: Int, id: Long, model: Any) -> Unit)? = null
) : RecyclerView.Adapter<ViewHolder>() {
    val mValues: SortedList<BangumiEntity> = SortedList(
        BangumiEntity::class.java,
        object : SortedListAdapterCallback<BangumiEntity?>(this) {
            override fun compare(o1: BangumiEntity?, o2: BangumiEntity?): Int = 0

            override fun areContentsTheSame(
                oldItem: BangumiEntity?,
                newItem: BangumiEntity?
            ): Boolean = oldItem?.id == newItem?.id

            override fun areItemsTheSame(item1: BangumiEntity?, item2: BangumiEntity?): Boolean =
                item1?.id == item2?.id

        }
    )

    init {
        mValues.addAll(values)
    }

    override fun getItemCount(): Int {
        return mValues.size()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_onairfragment, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        mValues[position]?.run {
            image?.run {
                holder.mImageView.loadUrl(this, mValues[position].cover_color)
            }
            holder.itemView.setOnClickListener {
                onItemClickListener?.invoke(it, position, 0, this)
            }
            holder.mTitleTextView.text = LanguageSwitchUtils.switchLanguageToJa(
                holder.itemView.context,
                name,
                nameCn
            )
            holder.mSummaryTextView.text = summary

            val dayInWeek = ArrarysResourceUtils.dayInWeek(
                holder.itemView.context,
                airWeekday.toInt()
            )
            val resultString = holder.itemView.context.getString(
                R.string.formatter_day_airdate,
                airDate,
                dayInWeek
            )
            holder.mEtcTextView.text = resultString
        }
    }
}

class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
    val mView: CardView
    val mTitleTextView: TextView
    val mSummaryTextView: TextView
    val mEtcTextView: TextView
    val mImageView: ImageView

    init {
        mView = view as CardView
        mTitleTextView = view.findViewById(R.id.item_title_textview)
        mImageView = view.findViewById(R.id.item_album)
        mSummaryTextView = view.findViewById(R.id.item_subtitle_textview)
        mEtcTextView = view.findViewById(R.id.item_etc_textview)
    }
}